File: java/Paranthesis.java
import java.util.Stack;

public class Paranthesis
{

    // Function to check if the expression is balanced
public static boolean isBalanced(String exp)
    {
        Stack<Character> stack = new Stack<>();
        for (char ch : exp.toCharArray())
        {
            if (ch == '(' || ch == '{' || ch == '[')
            {
                stack.push(ch);
            }
            else if (ch == ')' || ch == '}' || ch == ']')
            {
                if (stack.isEmpty())
                    return false;
                char top = stack.pop();
                if ((ch == ')' && top != '(') ||
                    (ch == '}' && top != '{') ||
                    (ch == ']' && top != '['))
                {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }

    // Driver code
public static void main(String[] args)
    {
        String exp = "[()]{}{[()()]()}";
        if (isBalanced(exp))
            System.out.println("balanced");
        else
            System.out.println("not balanced");
    }
}

File: python/PARANTHESIS.py

def isBalanced(x):
    stack = []

    brackets = {')': '(', '}': '{', ']': '['}
    
    for char in x:

        if char in brackets.values():
            stack.append(char)

        elif char in brackets.keys():

            if stack == [] or stack.pop() != brackets[char]:
                return False

    return stack == []


expression = "[()]{}{[()()]()}"
if isBalanced(expression):
    print("balanced")
else:
    print("not balanced")
